elastic/kibana
# [Synthetics] stream results back for project monitors

## Summary

After https://github.com/elastic/uptime/issues/475, the project monitors api became much slower, often resulting in timeouts on the synthetic-agent side for even a small number of monitors, sometimes as low as 5. This PR accounts for that by streaming updates back to the synthetics agent. The final update in the stream remains the same, and consists of a full summary of the created, updated, deleted, and failed monitors. 

EDIT: Added by @justinkambic:

## Testing this PR

It is possible to test the `push` command from Synthetics with only a Kibana/ES cluster. It's also possible to test this E2E, with some additional configuration.

### Testing `push` only

You can configure Kibana to do this in a few minutes.

1. Start up Kibana against a clean ES cluster
2. Navigate to Uptime
3. Enable Monitor Management
  <img width="1178" alt="image" src="https://user-images.githubusercontent.com/18429259/184381570-1cb75ffa-05ab-4c1b-af4c-faed1b2d25b7.png">
4. Click `Private Locations`, and follow the flow to make an Agent Policy
  <img width="1484" alt="image" src="https://user-images.githubusercontent.com/18429259/184381696-4956c8dd-0220-45d9-92d9-b9acf7b24a8b.png">
  <img width="680" alt="image" src="https://user-images.githubusercontent.com/18429259/184381773-2ed0e043-1e5c-47d0-a34b-a2a531fa61ac.png">
5. Navigate back to Monitor Management, follow the flow to add a Private Location
  <img width="543" alt="image" src="https://user-images.githubusercontent.com/18429259/184382030-3308259a-9c9e-4114-8a44-95c79159489a.png">
  <img width="531" alt="image" src="https://user-images.githubusercontent.com/18429259/184382094-cf402f47-82b8-428e-b40a-7a2c845799b7.png">
6. Create an [API key](https://www.elastic.co/guide/en/kibana/master/api-keys.html) for your admin user in Kibana.
7. Checkout the [synthetics branch that supports push streaming](https://github.com/elastic/synthetics/pull/576).
8. `npm run build && npm link`. Then, make a new directory for the project you will push. `mkdir {PROJECT_NAME}`. For the sake of clarity, make this project folder the same as the private location name you specified.
9. Use the build Synthetics CLI to create your project. `node dist/cli init {PROJECT_PATH}`. Follow the prompts to configure your project. You should see your private location on the list of choices in the CLI, select that one.
10. Enter your project dir and run `npm link @elastic/synthetics`.
11. After Synthetics builds your project, run `push` to see if your project's monitors get loaded to your Kibana instance. Example: `env SYNTHETICS_API_KEY={API_KEY} npm run push`. You should see output like the example below:

```
> preparing all monitors
> creating all monitors
> addition and completion of single task: monitor created successfully
> adding and removing multiple tasks: monitor created successfully
> example-monitor: monitor created successfully
✓ Pushed
```


### Testing this patch e2e

To do this, you will largely need to replicate the previous testing example, but with some additional configuration.

Firstly, set up the `elastic-package` utility.

0. Run `elastic-package stack up -d -v --version 8.5.0-SNAPSHOT` to spin up a deployment on 8.5 that we can use for testing purposes.
1. Modify your local Kibana to run on port 5602 with `server.port: 5602`. `elastic-package` also uses `https` by default now, so you'll need to modify your host config to account for this if you were using `http` previously. Example: `elasticsearch.hosts: https://localhost:9200`.
2. Before we can run Kibana, we need to update the `kibana_system` password for it.
```
curl -X POST https://localhost:9200/_security/user/kibana_system/_password\?pretty -u elastic:changeme -H 'Content-Type: application/json' -d '{"password": "changeme" }' --insecure
```
3. Start Kibana and log in, create a Private Location with the Elastic Agent policy.
  <img width="544" alt="image" src="https://user-images.githubusercontent.com/18429259/184390098-48f752f4-cfad-481a-9da7-a5c71304978f.png">
4. Define an API key for your user.
5. Follow the instructions in the previous steps to push monitors.
6. Navigate to Uptime, you should be able to observer your monitors running:
  <img width="1511" alt="image" src="https://user-images.githubusercontent.com/18429259/184392359-39b22cf2-1110-4cd5-9891-8b4238abfc28.png">



### Additional testing

- Try configuring project monitors for private locations in different spaces to ensure it's working.
- You can also create error states in Kibana's server by modifying the PR code.
